File: 005/solution_005.erl
-module(solution_005).
-export([main/0]).


main() ->
    project_euler_005(1, 20),
    halt().


project_euler_005(Bottom, Top) ->
    Solution = smallest_divisible(Bottom, Top),
    io:format("~b\n", [Solution]).


smallest_divisible(Bottom, Top) ->
    ByRange = lists:seq(Bottom, Top),
    Init = Top,
    Step = Top,
    smallest_divisible(Init, Step, ByRange).


smallest_divisible(N, Step, ByRange) ->
    case is_divisible(N, ByRange) of
        true -> N;
        false -> smallest_divisible(N + Step, Step, ByRange)
    end.


is_divisible(_, []) -> true;
is_divisible(N, [D | _Divisors]) when N rem D =/= 0 -> false;
is_divisible(N, [_ |  Divisors]) -> is_divisible(N, Divisors).

File: 005/solution_005.ml
let rec is_divisible = function
  | n, (bottom, top) when bottom > top -> true
  | n, (bottom, top) when n mod bottom != 0 -> false
  | n, (bottom, top) -> is_divisible (n, (bottom + 1, top))


let smallest_divisible (_bottom, top as by_range) =
  let rec smallest_divisible = function
    | n, step, by_range when is_divisible (n, by_range) -> n
    | n, step, by_range -> smallest_divisible (n + step, step, by_range)
  in
  let init = top
  and step = top in
  smallest_divisible (init, step, by_range)


let project_euler_005 by_range =
  let solution = smallest_divisible by_range in
  Printf.printf "%d\n" solution


let () = project_euler_005 (1, 20)
